python/pyspark/sql/types.py

```diff
@@ -168,10 +168,12 @@ def needConversion(self):
         return True
 
     def toInternal(self, d):
-        return d and d.toordinal() - self.EPOCH_ORDINAL
+        if d is not None:
```

Comment:
It will be false for None, empty string, empty list, 0. This way it will be returned instead of the date:
```
>>> [] and 'foo'
[]
>>> 0 and 'foo'
0
>>> 1 and 'foo'
'foo'
```